### ISSUE TITLE
failed stage resume: cannot find transaction 5b87d51c5540e3051751d249_0c5d64c0 in queue for document {actions 6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf}

### ISSUE THREAD

**Issue by freyes** [juju/mgopurge#28]
When running mgopurge in a long standing database used by juju 2.2.6 the following error is raised:

```
This program should only be used to recover from specific transaction
related problems in a Juju database. Casual use is strongly
discouraged. Irreversible damage may be caused to a Juju deployment
through improper use of this tool.

Aside from limited cases, this program should not be run while Juju
controller machine agents are running.

Ok to proceed? [y/n] y
2018-09-05 12:09:21 INFO  Running stage "presence": Clear presence data. This will be recreated when the controllers restart.
2018-09-05 12:09:21 INFO  Running stage "purgemissing": Purge orphaned transactions
2018-09-05 12:10:30 INFO  Running stage "trim": Trim txn-queues that are longer than 1000
2018-09-05 12:10:41 INFO  "units" document 6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/0 has 284298 tokens
2018-09-05 12:10:41 INFO  "units" document 6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/1 has 2379 tokens
2018-09-05 12:10:41 INFO  "units" document 6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/2 has 2374 tokens
2018-09-05 12:10:41 INFO  found 208106 transactions and 289051 tokens that might be trimmed
2018-09-05 12:11:01 DEBUG trim removed 10000 txns in 18.152s (skipped 0) from txns, pulled 0 tokens in 0.000s
2018-09-05 12:13:19 DEBUG trim removed 14788 txns in 18.311s (skipped 0) from txns, pulled 16163 tokens in 137.344s
2018-09-05 12:14:18 DEBUG trim removed 24788 txns in 37.059s (skipped 35212) from txns, pulled 16163 tokens in 137.344s
2018-09-05 12:17:45 DEBUG trim removed 33333 txns in 51.254s (skipped 35212) from txns, pulled 40418 tokens in 330.419s
2018-09-05 12:18:40 DEBUG trim removed 43333 txns in 65.723s (skipped 66667) from txns, pulled 40418 tokens in 330.419s
2018-09-05 12:18:57 DEBUG trim removed 53333 txns in 82.774s (skipped 66667) from txns, pulled 40418 tokens in 330.419s
2018-09-05 12:23:46 DEBUG trim removed 55196 txns in 82.959s (skipped 66667) from txns, pulled 74701 tokens in 618.999s
2018-09-05 12:24:36 DEBUG trim removed 65196 txns in 97.540s (skipped 94804) from txns, pulled 74701 tokens in 618.999s
2018-09-05 12:28:37 DEBUG trim removed 76087 txns in 112.385s (skipped 94804) from txns, pulled 108976 tokens in 844.442s
2018-09-05 12:28:46 INFO  trimmed 3 docs from 110748 tokens (852.739s), removing 77210 transactions (112.450s) in 18m15.210365125s
2018-09-05 12:28:46 INFO  Running stage "resume": Resume incomplete transactions
2018-09-05 12:28:46 ERROR failed stage resume: cannot find transaction 5b87d51c5540e3051751d249_0c5d64c0 in queue for document {actions 6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf}
```

The document with this id:

```
juju:PRIMARY> db.actions.find({"_id": "6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf"}).pretty();
{
	"_id" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf",
	"model-uuid" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de",
	"receiver" : "zookeeper/0",
	"name" : "juju-run",
	"parameters" : {
		"command" : "relation-get -r zkpeer:47 private-address zookeeper/0",
		"timeout" : NumberLong("300000000000")
	},
	"enqueued" : ISODate("2018-08-30T11:29:33Z"),
	"started" : ISODate("2018-08-30T11:29:36Z"),
	"completed" : ISODate("2018-08-30T11:29:36Z"),
	"status" : "completed",
	"message" : "",
	"results" : {
		"Code" : "0",
		"Stderr" : "",
		"Stdout" : "1.2.3.4\n"
	},
	"txn-revno" : NumberLong(4),
	"txn-queue" : [ ]
}
```
One of the problems with this is that jujud is spamming the machine-0.log with the following messages:

```
2018-09-05 10:05:22 ERROR juju.worker.dependency engine.go:551 "mgo-txn-resumer" manifold worker returned unexpected error: cannot resume transactions: cannot find transaction 5b87d51c5540e3051751d249_0c5d64c0 in queue for document {actions 6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf}
```

**Comment by tbaumann** [comment]
I can provide more logs if it helps

**Comment by jameinel** [comment]
First, I'd like to see what the actual transaction looks like:
 db.txns.find({"_id": ObjectId("5b87d51c5540e3051751d249")}).pretty()

(Note the transaction _id is the string that you see (called a token), without suffix with the '_' and as an ObjectId, not a string)

I would guess that the transaction is in state "2" which means it was supposed to have updated all documents involved with its token. But for some reason that constraint isn't being honored.

The easy "fix" for this is to 
db.txns.update({"_id": ObjectId("5b87d51c5540e3051751d249")}, {"$set": {"s": 1}, "$unset": {"n": 1}})

Which puts it back into "Preparing" state and means that it will ensure that the transaction is recorded in all the documents affected by it.

**Comment by tbaumann** [comment]
```
 juju:PRIMARY>  db.txns.find({"_id": ObjectId("5b87d51c5540e3051751d249")}).pretty()
{
	"_id" : ObjectId("5b87d51c5540e3051751d249"),
	"s" : 2,
	"o" : [
		{
			"c" : "units",
			"d" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/0",
			"a" : {
				"life" : {
					"$ne" : 2
				}
			}
		},
		{
			"c" : "actions",
			"d" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf",
			"a" : "d-",
			"i" : {
				"_id" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf",
				"model-uuid" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de",
				"receiver" : "zookeeper/0",
				"name" : "juju-run",
				"parameters" : {
					"command" : "relation-get -r zkpeer:47 private-address zookeeper/0",
					"timeout" : NumberLong("300000000000")
				},
				"enqueued" : ISODate("2018-08-30T11:29:33Z"),
				"started" : ISODate("0001-01-01T00:00:00Z"),
				"completed" : ISODate("0001-01-01T00:00:00Z"),
				"status" : "pending",
				"message" : "",
				"results" : {
					
				}
			}
		},
		{
			"c" : "actionnotifications",
			"d" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/0_a_eb08e238-dfa8-4c46-842e-47a4eb929adf",
			"a" : "d-",
			"i" : {
				"_id" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de:zookeeper/0_a_eb08e238-dfa8-4c46-842e-47a4eb929adf",
				"model-uuid" : "6a783ac4-0b48-45a3-87fc-9646f8bd82de",
				"receiver" : "zookeeper/0",
				"actionid" : "eb08e238-dfa8-4c46-842e-47a4eb929adf"
			}
		}
	],
	"n" : "0c5d64c0"
}
juju:PRIMARY> 

```

**Comment by tbaumann** [comment]
I stopped the controllers before I made the change.
Controllers started after that but were heavily loaded and unreliable responding to api calls.

What happened in the meantime since the issue was reported is that logs grew very much as well.
2018-09-09 20:50:05 ERROR juju.worker.dependency engine.go:551 "mgo-txn-resumer" manifold worker returned unexpected error: cannot resume transactions: cannot find transaction
5b87d51c5540e3051751d249_0c5d64c0 in queue for document {actions 6a783ac4-0b48-45a3-87fc-9646f8bd82de:eb08e238-dfa8-4c46-842e-47a4eb929adf}
2018-09-09 20:50:09 ERROR juju.worker.dependency engine.go:551 "mgo-txn-resumer" manifold worker returned unexpected error: cannot resume transactions: Resulting document after update is larger than 16777216

Also txn queue was massive as well.

Then I used mgopurge to prune some finished transactions as the txn queue ha grown to extreme size.
mgopurge resume would OOM.

Also I dropped a ton of logs, and since then the controllers immediately blow up in memory consumption like mgopurge resume did.
See https://bugs.launchpad.net/juju/+bug/1797816

**Comment by tbaumann** [comment]
5b87d51c5540e3051751d249 switched back to s='2'

**Comment by hloeung** [comment]
Ran into this earlier today, John's fix worked to unblock mgopurge.

    2020-11-22 23:27:52 ERROR failed stage resume: cannot find transaction 5fa48d53c2b5ab7d2eb82a78_6e078a81 in queue for document {relations b6be743c-3db6-4edb-83c2-ba125fb20dba:remote-57168d7f6bb24d9a8951a2b7465322f5:db
postgresql-...:db}
    db.txns.update({"_id": ObjectId("5fa48d53c2b5ab7d2eb82a78")}, {"$set": {"s": 1}, "$unset": {"n": 1}})

**Comment by hloeung** [comment]
Also, is this a duplicate of https://github.com/juju/mgopurge/issues/13 ?
